File: sirius-core/src/main/java/org/starcoin/sirius/core/NodeInformation.java
package org.starcoin.sirius.core;

import org.starcoin.core.Hash;
import org.starcoin.core.ProtobufCodec;
import org.starcoin.proto.Starcoin;

import java.util.Objects;

public class NodeInformation implements ProtobufCodec<Starcoin.ProtoNodeInfo> {

    public static final NodeInformation EMPTY_NODE = new NodeInformation(Hash.ZERO_HASH, 0,
            Hash.ZERO_HASH);

    private Hash left;
    private long offset;
    private Hash right;

    public NodeInformation() {
    }

    public NodeInformation(Hash left, long offset, Hash right) {
        this.left = left;
        this.offset = offset;
        this.right = right;
    }

    public Hash getLeft() {
        return left;
    }

    public long getOffset() {
        return offset;
    }

    public Hash getRight() {
        return right;
    }

    @Override
    public Starcoin.ProtoNodeInfo marshalProto() {
        return Starcoin.ProtoNodeInfo.newBuilder()
                .setLeft(this.left.toProto())
                .setOffset(this.offset)
                .setRight(this.right.toProto())
                .build();
    }

    @Override
    public void unmarshalProto(Starcoin.ProtoNodeInfo proto) {
        this.left = Hash.wrap(proto.getLeft());
        this.offset = proto.getOffset();
        this.right = Hash.wrap(proto.getRight());
    }

    public static NodeInformation generateNodeInformation(Starcoin.ProtoNodeInfo proto) {
        if (proto == null) {
            return null;
        }
        NodeInformation nodeInformation = new NodeInformation();
        nodeInformation.unmarshalProto(proto);
        return nodeInformation;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (!(o instanceof NodeInformation)) {
            return false;
        }
        NodeInformation that = (NodeInformation) o;
        return offset == that.offset &&
                Objects.equals(left, that.left) &&
                Objects.equals(right, that.right);
    }

    @Override
    public int hashCode() {
        return Objects.hash(left, offset, right);
    }
}

File: sirius-core/src/main/kotlin/org/starcoin/sirius/core/NodeInformation.kt
package org.starcoin.sirius.core

import kotlinx.serialization.SerialId
import kotlinx.serialization.Serializable
import org.apache.commons.lang3.RandomUtils
import org.starcoin.proto.Starcoin
import org.starcoin.sirius.serialization.ProtobufSchema

@ProtobufSchema(Starcoin.ProtoNodeInfo::class)
@Serializable
data class NodeInformation(
    @SerialId(1)
    val left: Hash = Hash.ZERO_HASH,
    @SerialId(2)
    val offset: Long = 0,
    @SerialId(3)
    val right: Hash = Hash.ZERO_HASH
) : SiriusObject() {
    companion object : SiriusObjectCompanion<NodeInformation, Starcoin.ProtoNodeInfo>(NodeInformation::class) {

        val EMPTY_NODE = NodeInformation(
            Hash.ZERO_HASH, 0,
            Hash.ZERO_HASH
        )

        override fun parseFromProtoMessage(proto: Starcoin.ProtoNodeInfo): NodeInformation {
            return NodeInformation(Hash.wrap(proto.left), proto.offset, Hash.wrap(proto.right))
        }

        override fun toProtoMessage(obj: NodeInformation): Starcoin.ProtoNodeInfo {
            return Starcoin.ProtoNodeInfo.newBuilder().setLeft(obj.left.toByteString()).setOffset(obj.offset)
                .setRight(obj.right.toByteString()).build()
        }

        @Deprecated("use parseFromProtoMessage", replaceWith = ReplaceWith("parseFromProtoMessage", ""))
        fun generateNodeInformation(proto: Starcoin.ProtoNodeInfo): NodeInformation? {
            return parseFromProtoMessage(proto)
        }

        override fun mock(): NodeInformation {
            return NodeInformation(Hash.random(), RandomUtils.nextLong(), Hash.random())
        }
    }
}
